File: BrewPOS-Workspace/brewpos-backend/src/routes/inventory.ts
import { Router } from 'express';
import { prisma } from '../db';
import multer from 'multer';
import path from 'path';
import fs from 'fs';

const router = Router();

// Setup Multer for Image Upload
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    const dir = path.join(__dirname, '../../public/uploads');
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
    cb(null, dir);
  },
  filename: (req, file, cb) => {
    const uniqueSuffix = Date.now() + '-' + Math.round(Math.random() * 1e9);
    cb(null, 'inventory-' + uniqueSuffix + path.extname(file.originalname));
  },
});
const upload = multer({ storage });

// GET all inventory items
router.get('/', async (req, res) => {
  try {
    const items = await prisma.inventoryItem.findMany();
    res.json(items);
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch inventory' });
  }
});

// POST new inventory item
router.post('/', upload.single('image'), async (req, res) => {
  try {
    const { name, stock, unit } = req.body;
    let imageUrl = null;
    if (req.file) {
      imageUrl = `/uploads/${req.file.filename}`;
    }

    const item = await prisma.inventoryItem.create({
      data: { 
        name, 
        stock: parseInt(stock) || 0, 
        unit: unit || 'pcs',
        imageUrl
      },
    });
    // Create initial log
    if (item.stock > 0) {
      await prisma.inventoryLog.create({
        data: { itemId: item.id, quantity: item.stock, type: 'IN', notes: 'Initial stock' }
      });
    }
    res.status(201).json(item);
  } catch (error) {
    res.status(500).json({ error: 'Failed to create inventory item' });
  }
});

// PUT edit inventory item (name, unit, image)
router.put('/:id', upload.single('image'), async (req, res) => {
  try {
    const { id } = req.params;
    const { name, unit } = req.body;

    const existing = await prisma.inventoryItem.findUnique({ where: { id: parseInt(id) } });
    if (!existing) return res.status(404).json({ error: 'Item not found' });

    let imageUrl = existing.imageUrl;
    if (req.file) {
      imageUrl = `/uploads/${req.file.filename}`;
      // Optional: Delete old image here if exists
    }

    const item = await prisma.inventoryItem.update({
      where: { id: parseInt(id) },
      data: { name, unit, imageUrl },
    });
    res.json(item);
  } catch (error) {
    res.status(500).json({ error: 'Failed to update inventory item' });
  }
});

// PUT update stock (adjust in/out)
router.put('/:id/adjust', async (req, res) => {
  try {
    const { id } = req.params;
    const { quantity, type, notes } = req.body; // type: "IN" or "OUT", quantity is absolute positive

    const item = await prisma.inventoryItem.findUnique({ where: { id: parseInt(id) } });
    if (!item) return res.status(404).json({ error: 'Item not found' });

    const q = parseInt(quantity);
    if (isNaN(q) || q <= 0) return res.status(400).json({ error: 'Invalid quantity' });

    const newStock = type === 'IN' ? item.stock + q : item.stock - q;
    
    const updated = await prisma.inventoryItem.update({
      where: { id: parseInt(id) },
      data: { stock: newStock }
    });

    await prisma.inventoryLog.create({
      data: { itemId: parseInt(id), quantity: q, type, notes }
    });

    res.json(updated);
  } catch (error) {
    res.status(500).json({ error: 'Failed to adjust inventory stock' });
  }
});

// GET /opname/today
router.get('/opname/today', async (req, res) => {
  try {
    const opname = await prisma.dailyOpname.findFirst({
      where: { status: 'OPEN' },
      include: { items: { include: { inventoryItem: true } } },
      orderBy: { createdAt: 'desc' }
    });
    
    if (opname) {
      // Dynamically compute addedStock for UI display before closing
      for (let opItem of opname.items) {
        const logs = await prisma.inventoryLog.aggregate({
          where: {
            itemId: opItem.inventoryItemId,
            type: 'IN',
            createdAt: { gte: opname.createdAt }
          },
          _sum: { quantity: true }
        });
        (opItem as any).addedStock = logs._sum.quantity || 0;
      }
      return res.json({ status: 'OPEN', opname });
    }

    const inventory = await prisma.inventoryItem.findMany();
    res.json({ status: 'NOT_OPEN', inventory });
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch opname status' });
  }
});

// POST /opname/open
router.post('/opname/open', async (req, res) => {
  try {
    const { items } = req.body;
    
    const existing = await prisma.dailyOpname.findFirst({ where: { status: 'OPEN' } });
    if (existing) return res.status(400).json({ error: 'Opname already open' });

    const opname = await prisma.dailyOpname.create({
      data: {
        status: 'OPEN',
        items: {
          create: items.map((i: any) => ({
            inventoryItemId: i.inventoryItemId,
            openingStock: i.openingStock,
            morningNotes: i.morningNotes,
          }))
        }
      },
      include: { items: true }
    });

    for (const i of items) {
      await prisma.inventoryItem.update({
        where: { id: i.inventoryItemId },
        data: { stock: i.openingStock }
      });
    }

    res.status(201).json(opname);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to open opname' });
  }
});

// POST /opname/confirm-stock
router.post('/opname/confirm-stock', async (req, res) => {
  try {
    const opname = await prisma.dailyOpname.findFirst({
      where: { status: 'OPEN' }
    });
    if (!opname) {
      return res.status(400).json({ error: 'No open opname found' });
    }

    const updatedOpname = await prisma.dailyOpname.update({
      where: { id: opname.id },
      data: { isStockConfirmed: true }
    });

    res.json(updatedOpname);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to confirm stock' });
  }
});

// POST /opname/cancel-confirm-stock
router.post('/opname/cancel-confirm-stock', async (req, res) => {
  try {
    const opname = await prisma.dailyOpname.findFirst({
      where: { status: 'OPEN' }
    });
    if (!opname) {
      return res.status(400).json({ error: 'No open opname found' });
    }

    const updatedOpname = await prisma.dailyOpname.update({
      where: { id: opname.id },
      data: { isStockConfirmed: false }
    });

    res.json(updatedOpname);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to cancel confirm stock' });
  }
});

// POST /opname/close
router.post('/opname/close', async (req, res) => {
  try {
    const { opnameId, items } = req.body;
    
    const opname = await prisma.dailyOpname.findUnique({ where: { id: opnameId }, include: { items: true } });
    if (!opname || opname.status === 'CLOSED') {
      return res.status(400).json({ error: 'Invalid or already closed opname' });
    }

    for (const i of items) {
      const opItem = opname.items.find((oi: any) => oi.inventoryItemId === i.inventoryItemId);
      if (opItem) {
        // Find added stock during this opname period
        const logs = await prisma.inventoryLog.aggregate({
          where: {
            itemId: i.inventoryItemId,
            type: 'IN',
            createdAt: { gte: opname.createdAt }
          },
          _sum: { quantity: true }
        });
        const addedStock = logs._sum.quantity || 0;
        const used = (opItem.openingStock + addedStock) - i.closingStock;

        await prisma.dailyOpnameItem.update({
          where: { id: opItem.id },
          data: {
            closingStock: i.closingStock,
            addedStock: addedStock,
            used: used,
            nightNotes: i.nightNotes
          }
        });

        await prisma.inventoryItem.update({
          where: { id: i.inventoryItemId },
          data: { stock: i.closingStock }
        });
      }
    }

    const updatedOpname = await prisma.dailyOpname.update({
      where: { id: opnameId },
      data: { status: 'CLOSED' },
      include: { items: { include: { inventoryItem: true } } }
    });

    res.json(updatedOpname);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to close opname' });
  }
});

// GET /opname/history
router.get('/opname/history', async (req, res) => {
  try {
    const history = await prisma.dailyOpname.findMany({
      where: { status: 'CLOSED' },
      include: { items: { include: { inventoryItem: true } } },
      orderBy: { createdAt: 'desc' }
    });
    res.json(history);
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch opname history' });
  }
});

export default router;

File: BrewPOS-Workspace/brewpos-flask/app/models/inventory.py
from datetime import datetime
from app.extensions import db

class InventoryItem(db.Model):
    __tablename__ = 'InventoryItem'

    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    name = db.Column(db.String(255), unique=True, nullable=False)
    stock = db.Column(db.Float, nullable=False, default=0.0)
    unit = db.Column(db.String(50), nullable=False, default='pcs')  # pcs, gram, ml, shot, bag, etc.
    minStock = db.Column(db.Float, nullable=False, default=10.0)
    costPerUnit = db.Column(db.Float, nullable=False, default=0.0)  # Purchase cost per 1 unit (gram, ml, pcs)
    imageUrl = db.Column(db.String(500), nullable=True)

    logs = db.relationship('InventoryLog', backref='item', lazy=True, cascade='all, delete-orphan')
    opnameItems = db.relationship('DailyOpnameItem', backref='inventoryItem', lazy=True)

    def to_dict(self):
        # If stock is a whole number, format cleanly or keep float
        stock_val = int(self.stock) if self.stock is not None and self.stock.is_integer() else (self.stock or 0)
        min_stock_val = int(self.minStock) if self.minStock is not None and self.minStock.is_integer() else (self.minStock or 10.0)
        return {
            'id': self.id,
            'name': self.name,
            'stock': stock_val,
            'unit': self.unit,
            'minStock': min_stock_val,
            'costPerUnit': self.costPerUnit or 0.0,
            'imageUrl': self.imageUrl
        }


class InventoryLog(db.Model):
    __tablename__ = 'InventoryLog'

    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    itemId = db.Column(db.Integer, db.ForeignKey('InventoryItem.id', ondelete='CASCADE'), nullable=False)
    quantity = db.Column(db.Float, nullable=False)
    type = db.Column(db.String(20), nullable=False)  # 'IN' or 'OUT'
    totalCost = db.Column(db.Float, nullable=False, default=0.0)
    costPerUnit = db.Column(db.Float, nullable=False, default=0.0)
    supplier = db.Column(db.String(255), nullable=True)
    notes = db.Column(db.Text, nullable=True)
    receiptUrl = db.Column(db.String(500), nullable=True)
    createdAt = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)

    def to_dict(self):
        qty_val = int(self.quantity) if self.quantity is not None and self.quantity.is_integer() else (self.quantity or 0)
        return {
            'id': self.id,
            'itemId': self.itemId,
            'quantity': qty_val,
            'type': self.type,
            'totalCost': self.totalCost or 0.0,
            'costPerUnit': self.costPerUnit or 0.0,
            'supplier': self.supplier,
            'notes': self.notes,
            'receiptUrl': self.receiptUrl,
            'createdAt': self.createdAt.isoformat() if self.createdAt else None
        }


class DailyOpname(db.Model):
    __tablename__ = 'DailyOpname'

    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    date = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)
    status = db.Column(db.String(20), nullable=False, default='OPEN')  # OPEN, CLOSED
    isStockConfirmed = db.Column(db.Boolean, nullable=False, default=False)
    createdAt = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)
    updatedAt = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    items = db.relationship('DailyOpnameItem', backref='opname', lazy=True, cascade='all, delete-orphan')

    def to_dict(self):
        return {
            'id': self.id,
            'date': self.date.isoformat() if self.date else None,
            'status': self.status,
            'isStockConfirmed': self.isStockConfirmed,
            'items': [item.to_dict() for item in self.items],
            'createdAt': self.createdAt.isoformat() if self.createdAt else None,
            'updatedAt': self.updatedAt.isoformat() if self.updatedAt else None
        }


class DailyOpnameItem(db.Model):
    __tablename__ = 'DailyOpnameItem'

    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    opnameId = db.Column(db.Integer, db.ForeignKey('DailyOpname.id', ondelete='CASCADE'), nullable=False)
    inventoryItemId = db.Column(db.Integer, db.ForeignKey('InventoryItem.id'), nullable=False)
    openingStock = db.Column(db.Float, nullable=False, default=0.0)
    addedStock = db.Column(db.Float, nullable=False, default=0.0)
    closingStock = db.Column(db.Float, nullable=True)
    used = db.Column(db.Float, nullable=True)
    morningNotes = db.Column(db.Text, nullable=True)
    nightNotes = db.Column(db.Text, nullable=True)

    def to_dict(self):
        def _clean_num(val):
            if val is None:
                return None
            return int(val) if isinstance(val, (int, float)) and float(val).is_integer() else val

        return {
            'id': self.id,
            'opnameId': self.opnameId,
            'inventoryItemId': self.inventoryItemId,
            'openingStock': _clean_num(self.openingStock),
            'addedStock': _clean_num(self.addedStock),
            'closingStock': _clean_num(self.closingStock),
            'used': _clean_num(self.used),
            'morningNotes': self.morningNotes,
            'nightNotes': self.nightNotes,
            'inventoryItem': self.inventoryItem.to_dict() if self.inventoryItem else None
        }
